cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicaProvider.java

```diff
@@ -0,0 +1,165 @@
+
+import java.math.BigInteger;
+import java.net.UnknownHostException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.sidecar.common.JmxClient;
+import org.apache.cassandra.sidecar.common.data.TokenRangeReplicasResponse;
+import org.jetbrains.annotations.NotNull;
+
+import static org.apache.cassandra.sidecar.cassandra40.EndpointSnitchJmxOperations.ENDPOINT_SNITCH_INFO_OBJ_NAME;
+import static org.apache.cassandra.sidecar.cassandra40.StorageJmxOperations.STORAGE_SERVICE_OBJ_NAME;
+
+/**
+ * Aggregates the replica-set by token range
+ */
+public class TokenRangeReplicaProvider
+{
+    private final JmxClient jmxClient;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TokenRangeReplicaProvider.class);
+
+    public TokenRangeReplicaProvider(JmxClient jmxClient)
+    {
+        this.jmxClient = jmxClient;
+    }
+
+    public TokenRangeReplicasResponse tokenRangeReplicas(String keyspace, Partitioner partitioner)
+    {
+        Objects.requireNonNull(keyspace, "keyspace must be non-null");
+
+        StorageJmxOperations storage = jmxClient.proxy(StorageJmxOperations.class, STORAGE_SERVICE_OBJ_NAME);
+
+        // Retrieve map of primary token ranges to endpoints that describe the ring topology
+        Map<List<String>, List<String>> rangeToEndpointMappings = storage.getRangeToEndpointWithPortMap(keyspace);
+        // Pending ranges include bootstrap tokens and leaving endpoints as represented in the Cassandra TokenMetadata
+        Map<List<String>, List<String>> pendingRangeMappings = storage.getPendingRangeToEndpointWithPortMap(keyspace);
+
+        Stream<String> hostsStream = Stream.concat(rangeToEndpointMappings.values().stream().flatMap(List::stream),
+                                                   pendingRangeMappings.values().stream().flatMap(List::stream));
+
+        Map<String, String> hostToDatacenter = groupHostsByDatacenter(hostsStream);
+
+        // Retrieve map of all token ranges (pending & primary) to endpoints
+        List<TokenRangeReplicasResponse.ReplicaInfo> writeReplicas =
+        writeReplicasFromPendingRanges(rangeToEndpointMappings,
+                                       pendingRangeMappings,
+                                       hostToDatacenter,
+                                       partitioner,
+                                       keyspace);
+
+        return new TokenRangeReplicasResponse(
+                writeReplicas,
+                mappingsToUnwrappedReplicaSet(rangeToEndpointMappings, hostToDatacenter, partitioner));
+    }
+
+    private List<TokenRangeReplicasResponse.ReplicaInfo>
+    writeReplicasFromPendingRanges(Map<List<String>, List<String>> readReplicaMappings,
```

Comment:
It is confusing to call it read replicas. We can use the same term, natural, as used in Cassandra code base.
```suggestion
    writeReplicasFromPendingRanges(Map<List<String>, List<String>> naturalReplicaMappings,
```

Reply:
yeah, makes sense, it always felt odd to call it read replicas.